### PR TITLE
[SPARK-21923][CORE]Avoid calling reserveUnrollMemoryForThisTask for every record

### PR DESCRIPTION
**Pull request by ConeyLiu**
## What changes were proposed in this pull request?
When Spark persist data to Unsafe memory, we call  the method `MemoryStore.putIteratorAsBytes`, which need synchronize the `memoryManager` for every record write. This implementation is not necessary, we can apply for more memory at a time to reduce unnecessary synchronization.

## How was this patch tested?

Test case (with 1 executor 20 core):
```scala
val start = System.currentTimeMillis()
val data = sc.parallelize(0 until Integer.MAX_VALUE, 100)
      .persist(StorageLevel.OFF_HEAP)
      .count()

println(System.currentTimeMillis() - start)

```

Test result:

before

|  27647  |  29108  |  28591  |  28264  |  27232  |

after

|  26868  |  26358  |  27767  |  26653  |  26693  |

### CODE REVIEW

**Comment by ConeyLiu** [comment]
Hi, @cloud-fan @jerryshao Would you mind take a look ? Thanks a lot

**Comment by ConeyLiu** [comment]
Here https://github.com/apache/spark/pull/19135/files?diff=unified#diff-870cd3693df7a5add2ac3119d7d91d34L373, we call `reserveAdditionalMemoryIfNecessary()` for every record write.

**Comment by jerryshao** [comment]
Sorry I'm not so familiar with this part, but from the test result seems that the performance just improved a little. I would doubt the way you generate RDD `0 until Integer.MAX_VALUE` might take most of the time (since a large integer array needs to be serialized with tasks and ship to executor). 

Also I see you use 1 executor with 20 cores to do test. In the normal usage case we will not allocate so many cores to 1 executor, can you please test with 2-4 cores per executor, I guess with less cores, the contention of MemoryManager lock should be alleviated, and the performance might be close.

**Comment by ConeyLiu** [comment]
Firstly, Serialization time did not take a long time. You can see follow: 
untitled

Secondly, I do not think that every executor in a distributed system should be set to very little core and memory. Because the more the process also means that more communication between the process, which means more data serialization and deserialization.

Thirdly, only when there are enough concurrent threads, thread synchronization will cause performance problems. In the server, we have 70 to 80 cores, concurrent tasks more than this.

This change is really small, the proportion of the entire task is also very small, so the impact on the total time is not so big, but in this test case, still increased by 5%.

**Comment by cloud-fan** [comment]
Does this patch has regressions? It seems to me that allocating more memory may starve other tasks/operators and reduce the overall performance.

**Comment by ConeyLiu** [comment]
hi @cloud-fan, The previous writing is the same as `putIteratorAsValues`. Now I have modified the code, each application for an additional `chunkSize` bytes of memory, because the size of `ChunkedByteBufferOutputStream` each growth is just `chunkSize`.

**Comment by jiangxb1987** [comment]
It would be great to test the perf on executors with various corsPerExecutor settings to ensure we don't bring in regressions by the code change.

**Comment by ConeyLiu** [comment]
@jiangxb1987 Ok, I can test it later. The following picture is when I run kmeans and put the source data into the offheap memory, and you can see the CPU time occupied by `reserveUnrollMemoryForThisTask` is very high.
pic

**Comment by cloud-fan** [comment]
is it better to do batch unrolling? i.e., we can check memory usage and request memory for like every 10 records, instead of doing it for every record.

**Comment by ConeyLiu** [comment]
@cloud-fan Very sorry to reply so late. I updated the code followed your suggestion. Does this need performance test? If needed, I will test it late.

**Comment by ConeyLiu** [comment]
Hi @cloud-fan , @jiangxb1987 . The follow is the test result, I have modified the test to server.
2 cores with 150GB memory:

| ------| ------ | ------ |
| before | after | proportion|
| 135881| 135183|  |
| 139352| 135652|  |
| 138356| 136431|  |
| 137041| 137980|  |
| 137191| 139708|  |
| 137564.2 | 136990.8 | 1.004186 | 

20 cores with 150GB memory:

| ------| ------ | ------ |
| before | after | proportion|
| 27519| 27683|  |
| 27319| 28055|  |
| 25905| 27357|  |
| 26362| 28709|  |
| 25042| 29306|  |
| 26429.4| 28222| 0.936482| 

50 cores with 150GB memory

| ------| ------ | ------ |
| before | after | proportion|
| 29740| 30570|  |
| 28573| 29881|  |
| 28922| 30221|  |
| 29809| 30096|  |
| 29069| 30349|  |
| 29222.6| 30223.4| 0.966887|

**Comment by jerryshao** [comment]
So it somehow reflects that CPU core contention is the main issue for memory pre-occupation , am I right?

AFAIK from our customer, we usually don't allocate so many cores to one executor, also avoid big executor (many cores + large heap memory) to reduce GC and other memory/cpu contentions.

**Comment by ConeyLiu** [comment]
Hi @jerryshao thanks for your reviewing.

>So it somehow reflects that CPU core contention is the main issue for memory pre-occupation
I have modified the code, now it will not request more memory, now it just reduce the times of calling `reserveUnrollMemoryForThisTask` followed by @cloud-fan comments. And also the method is same as `putIteratorAsValues`.

Yeah, its impact will be small with small cores. In the above test results,  it doesn't bring any regressions, and also better for many cores.  For machine learning, we need cache the source data to OFF_HEAP in order to reduce the gc problem.

For the configuration, I think the different application scenarios may be different.

File: core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala [hunk 1]
```diff
@@ -325,6 +325,10 @@ private[spark] class MemoryStore(
 
     // Whether there is still enough memory for us to continue unrolling this block
     var keepUnrolling = true
+    // Number of elements unrolled so far
+    var elementsUnrolled = 0L
+    // How often to check whether we need to request more memory
+    val memoryCheckPeriod = 16
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala]: This is the same approach we used in `putIteratorAsValues`. One thing still missing is `memoryGrowthFactor`, can we add that too?

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala]: in the future we should do some refactor between `putIteratorAsValues` and `putIteratorAsBytes`, to reduce duplicated code.

Comment (ConeyLiu) [reply core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala]: OK, I'll add it. I think I can take a try to refactor it. Do you need a separate pr?

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala]: yea definitely not in this PR.

Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala]: Should this be made configurable? Or can we make sure it's the best magic number that we can choose?

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala]: it's hard coded in `putIteratorAsValues` too, we can improve it later.

Comment (ConeyLiu) [reply core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala]: I have just made it configurable. I'm not sure if this writting is reasonable.

File: core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala [hunk 2]
```diff
@@ -190,11 +190,11 @@ private[spark] class MemoryStore(
     // Initial per-task memory to request for unrolling blocks (bytes).
     val initialMemoryThreshold = unrollMemoryThreshold
     // How often to check whether we need to request more memory
-    val memoryCheckPeriod = 16
+    val memoryCheckPeriod = conf.getLong("spark.storage.unrollMemoryCheckPeriod", 16)
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala]: we should move these 2 configs to `org.apache.spark.internal.config` and add some documents. They should be internal config I think.

Comment (ConeyLiu) [reply core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala]: OK, moved. Pls take a took.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 3]
```diff
@@ -385,4 +385,17 @@ package object config {
       .checkValue(v => v > 0 && v <= Int.MaxValue,
         s"The buffer size must be greater than 0 and less than ${Int.MaxValue}.")
       .createWithDefault(1024 * 1024)
+
+  private[spark] val UNROLL_MEMORY_CHECK_PERIOD =
+    ConfigBuilder("spark.storage.unrollMemoryCheckPeriod")
+      .doc("The memory check period is used to determine how often we should check whether "
+        + "there is a need to request more memory when we try to put the given block in memory.")
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala]: `put` -> `unroll`

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
LGTM except one minor comment

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 4]
```diff
@@ -385,4 +385,17 @@ package object config {
       .checkValue(v => v > 0 && v <= Int.MaxValue,
         s"The buffer size must be greater than 0 and less than ${Int.MaxValue}.")
       .createWithDefault(1024 * 1024)
+
+  private[spark] val UNROLL_MEMORY_CHECK_PERIOD =
+    ConfigBuilder("spark.storage.unrollMemoryCheckPeriod")
+      .doc("The memory check period is used to determine how often we should check whether "
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala]: let's call `.internal()` here, to make them internal configs, as we don't expect users to tune it.

**Comment by ConeyLiu** [comment]
Hi @cloud-fan, thanks for reviewing. The code has updated, pls take a look.

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**